arch/xtensa/soc/esp32/esp32-mp.c

```diff
@@ -0,0 +1,190 @@
+ * signal for "top of stack" to the debugger).
+ */
+void _appcpu_stack_switch(void *stack, void *entry);
+__asm__("\n"
+        ".align 4"		"\n"
+        "_appcpu_stack_switch:"	"\n\t"
+
+        "entry a1, 16"		"\n\t"
+
+	/* Subtle: we want the stack to be 16 bytes higher than the
+	 * top on entry to the called function, because the ABI forces
+	 * it to assume that those bytes are for its caller's A0-A3
+	 * spill area.  (In fact ENTRY instructions with stack
+	 * adjustments less than 16 are a warning condition in the
+	 * assembler). But we aren't a caller, have no bit set in
+	 * WINDOWSTART and will never be asked to spill anything.
+	 * Those 16 bytes would otherwise be wasted on the stack, so
+	 * adjust */
+	"addi a1, a2, 16"	"\n\t"
+
+        /* Clear WINDOWSTART so called functions never try to spill
+         * our callers' registers into the now-garbage stack pointers
+         * they contain.  No need to set the bit corresponding to
+         * WINDOWBASE, our C callee will do that when it does an
+         * ENTRY. */
+        "movi a0, 0"		"\n\t"
+        "wsr.WINDOWSTART a0"	"\n\t"
+
+        /* Clear CALLINC field of PS (you would think it would, but
+         * our ENTRY doesn't actually do that) so the callee's ENTRY
+         * doesn't shift the registers */
+        "rsr.PS a0"		"\n\t"
+        "movi a2, 0xfffcffff"	"\n\t"
+        "and a0, a0, a2"	"\n\t"
+        "wsr.PS a0"		"\n\t"
+
+        "rsync"			"\n\t"
+        "movi a0, 0"		"\n\t"
+
+	"jx a3"			"\n\t");
+
+/* Carefully constructed to use no stack beyond compiler-generated ABI
+ * instructions.  WE DO NOT KNOW WHERE THE STACK FOR THIS FUNCTION IS.
+ * The ROM library just picks a spot on its own with no input from our
+ * app linkage and tells us nothing about it until we're already
+ * running.
+ */
+static void appcpu_entry1(void)
+{
+	_appcpu_stack_switch(appcpu_top, appcpu_entry2);
+}
+
+/* The calls and sequencing here were extracted from the ESP-32
+ * FreeRTOS integration with just a tiny bit of cleanup.  None of the
+ * calls or registers shown are documented, so treat this code with
+ * extreme caution.
+ */
+static void appcpu_start(void)
+{
+	/* These two calls are wrapped in a "stall_other_cpu" API in
+	 * esp-idf.  But in this context the appcpu is stalled by
+	 * definition, so we can skip that complexity and just call
+	 * the ROM directly.
+	 */
+	esp32rom_Cache_Flush(1);
+	esp32rom_Cache_Read_Enable(1);
+
+	RTC_CNTL_SW_CPU_STALL &= ~RTC_CNTL_SW_STALL_APPCPU_C1;
+	RTC_CNTL_OPTIONS0     &= ~RTC_CNTL_SW_STALL_APPCPU_C0;
+	DPORT_APPCPU_CTRL_B   |= DPORT_APPCPU_CLKGATE_EN;
+	DPORT_APPCPU_CTRL_C   &= ~DPORT_APPCPU_RUNSTALL;
+
+	/* Pulse the RESETTING bit */
+	DPORT_APPCPU_CTRL_A |= DPORT_APPCPU_RESETTING;
+	DPORT_APPCPU_CTRL_A &= ~DPORT_APPCPU_RESETTING;
+
+	/* Seems weird that you set the boot address AFTER starting
+	 * the CPU, but this is how they do it...
+	 */
+	esp32rom_ets_set_appcpu_boot_addr((uint32_t)appcpu_entry1);
```

Comment:
Curious: does this work if you set the address before starting the CPU?

I wouldn't be surprised if it didn't work.  While working on enabling the flash cache, which touches some of these registers, there were some workarounds in the FreeRTOS port for HW bugs.